### DISCUSSION TITLE
(Done) January 8th 2025 Meetup: Clojure Native round-table

### DISCUSSION THREAD

**Discussion by frenchy64** [madclj/madclj.com#10] [Announcements]
👍 to RSVP

https://madclj.com/

Wednesday, January 8th, 2025
6:30 PM to 9:30 PM CT

[StartingBlock Madison](https://www.google.com/maps/search/?api=1&query=43.081245%2C%20-89.374054)
821 E Washington Ave 2nd floor · Madison, WI

Clojure was designed to be hosted on virtual machines like Java and CLR where
low-level concerns are handled by the virtual machine. This week our meetup
is all about connecting Clojure with native code and executables.

Take a look at the suggested topics below. If you'd like to lead or participate
in a discussion about them, follow some of the links and prepare brief presentation
(any format). We can have multiple presenters on the same topic, feel free to coordinate
in the GitHub Discussion.

Please suggest any related topics you'd like for us to cover or just bring them up
during the meetup.

Topics:

- Original design of Clojure on JVM
  - advantages + disadvantages
- Compiling Clojure to native executables with GraalVM
  - differences from JVM (Ambrose)
- Babashka + SCI
  - how is startup so fast
  - what's involved for library compatibility (Ambrose, James)
- Jank on LLVM
  - LLVM and how Jank generates [LLVM IR](https://llvm.org/docs/LangRef.html) (Sayan)
- Clang
  - how clang is relevant to Clojure
  - what are clang plugins
- JNI vs JNA vs FFI (new in JVM 22, aka Project Panama)
  - (James)

Links:

- [clong](https://github.com/phronmophobic/clong)
  - A wrapper for libclang and a generator that can turn c header files into clojure apis.
- [babashka](https://github.com/babashka/babashka)
  - Babashka is a native Clojure interpreter for scripting with fast startup.
- [SCI](https://github.com/babashka/SCI)
  - Clojure interpreter compilable with GraalVM native
- [clj-easy/graal-docs](https://github.com/clj-easy/graal-docs)
  - scripts and tips on natively compiling Clojure programs with GraalVM.
- [jank](https://github.com/jank-lang/jank)
  - jank is a Clojure dialect on LLVM with C++ interop.
- [jeps/454](https://openjdk.org/jeps/454)
  - an API by which Java programs can interoperate with code and data outside of the Java runtime.
- [jextract](https://github.com/openjdk/jextract)
  - jextract is a tool which mechanically generates Java bindings from native library headers.
  - via clang C API
- [vybe](https://github.com/pfeodrippe/vybe)
  - Clojure wrapper using Java's new Foreign Function & Memory API (JVM 22)
- [vybe.c](https://github.com/pfeodrippe/vybe/blob/develop/src/vybe/c.clj)
  - C compiler written in Clojure to extend Supercollider (Overtone)
  - [commentary](https://clojurians.slack.com/archives/C053Q1QSG/p1732687749027129?thread_ts=1731877266.734569&cid=C053Q1QSG)
  - [example usage](https://github.com/pfeodrippe/vybe/blob/641c00d07606c4cae126cee02930026e3d055f7f/src/vybe/experimental/audio/overtone.clj#L308)
- [tools-deps-native](https://github.com/babashka/tools-deps-native)
  - Clojure CLI on GraalVM

Agenda:

- 6:30pm arrive, food & drink provided
- 7pm, start structured discussion
- 8:30pm end structured discussion
- 9:30pm end meetup

**Comment by frenchy64** [comment]
I can talk about GraalVM and bb.

**Comment by frenchy64** [comment]
@madclj/meetup Topic for January 8th meetup is a round-table discussion about "Native Clojure". [RSVP](https://github.com/orgs/madclj/discussions/10)

**Comment by brunchboy** [comment]
I can talk briefly about how I tweaked my Beat Link library so that it could be used from GraalVM (for another project, since my own use of it is very JVM-Clojure focused, as I rely on `eval` being available, and make heavy use of a lot of Java libraries). I’d love to know if FFI is going to be convenient from Clojure, I may have to try to investigate that a bit before the meetup.

**Comment by sivakusayan** [comment]
I can talk briefly about LLVM and look into how Jank generates [LLVM IR](https://llvm.org/docs/LangRef.html)! 

There is also rich tooling that LLVM provides us to observe how it makes its optimization decisions, and I would love to (briefly) share the ones I've found so far 🙂

**Comment by dpetranek** [comment]
I'm going to be out of town on the 8th, have a good time and I'll catch you all at the next one!

**Comment by djwhitt** [comment]
Work once again prevents me. Catch you next time (hopefully)!

**Comment by frenchy64** [comment]
I've ordered pizzas and will be at the side door between the two buildings to let people in around 6:30pm.

**Comment by ericnormand** [comment]
Will be there in about 20 minutes!

**Comment by frenchy64** [comment]
![signal-2025-01-08-232745_002](https://github.com/user-attachments/assets/d5a04f81-9be1-4c3d-aaf9-b58ec01da709)

**Reply by frenchy64** [reply]
[Will there be a native version of Clojure in the future?](https://clojure.org/guides/faq#native)

https://llvm.godbolt.org/

https://github.com/sivakusayan/clang-plugin-boilerplate
